### ISSUE TITLE
async closures are incorrectly marked as not covered

### ISSUE THREAD

**Issue by lukaslihotzki-f** [rust-lang/rust#151135]
I tried this code:

```rust
// axum = "0.8"
// reqwest = "0.12"
// tokio = "1.47"

use axum::{Router, routing::get};

fn router() -> Router {
	Router::new().route(
		"/",
		get(async || {
			println!("Called async closure!");
		}),
	)
}

#[cfg(test)]
mod tests {
	use tokio::net::TcpListener;

	#[tokio::test]
	async fn test() {
		let listener = TcpListener::bind((std::net::Ipv6Addr::LOCALHOST, 0)).await.unwrap();
		let local_addr = listener.local_addr().unwrap();
		tokio::spawn(async { axum::serve(listener, crate::router()).await });
		reqwest::get(&format!("http://{}", local_addr)).await.unwrap();
	}
}
```

And captured the coverage with this command: `cargo llvm-cov nextest --lcov --output-path lcov.info --no-capture && genhtml lcov.info -o html`

I expected to see this happen: The `println!` is expected to be covered, because the output contains `Called async closure!`.

Instead, this happened: The `println!` line is marked as not covered.

### Meta

`rustc --version --verbose`:
```
rustc 1.92.0 (ded5c06cf 2025-12-08)
binary: rustc
commit-hash: ded5c06cf21d2b93bffd5d884aa6e96934ee4234
commit-date: 2025-12-08
host: x86_64-unknown-linux-gnu
release: 1.92.0
LLVM version: 21.1.3
```

The issue also exists in:
```
rustc 1.94.0-nightly (2850ca829 2026-01-13)
binary: rustc
commit-hash: 2850ca8295bc253186b291314ddc239632755475
commit-date: 2026-01-13
host: x86_64-unknown-linux-gnu
release: 1.94.0-nightly
LLVM version: 21.1.8
```

<details><summary>Backtrace (when replacing the `println` with `panic`)</summary>
<p>

```
Called async closure!
stack backtrace:
   0: __rustc::rust_begin_unwind
             at /rustc/ded5c06cf21d2b93bffd5d884aa6e96934ee4234/library/std/src/panicking.rs:698:5
   1: core::panicking::panic_fmt
             at /rustc/ded5c06cf21d2b93bffd5d884aa6e96934ee4234/library/core/src/panicking.rs:80:14
   2: exex::router::{closure#0}
             at ./src/lib.rs:7:4
   3: <exex::router::{closure#0} as axum::handler::Handler<((),), ()>>::call::{closure#0}
             at /home/lukas/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/axum-0.8.8/src/handler/mod.rs:216:38
   4: <core::pin::Pin<alloc::boxed::Box<dyn core::future::future::Future<Output = http::response::Response<axum_core::body::Body>> + core::marker::Send>> as core::future::future::Future>::poll
             at /home/lukas/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/future/future.rs:133:9
   5: <futures_util::future::future::map::Map<core::pin::Pin<alloc::boxed::Box<dyn core::future::future::Future<Output = http::response::Response<axum_core::body::Body>> + core::marker::Send>>, fn(http::response::Response<axum_core::body::Body>) -> core::result::Result<http::response::Response<axum_core::body::Body>, core::convert::Infallible>> as core::future::future::Future>::poll
             at /home/lukas/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/futures-util-0.3.31/src/future/future/map.rs:55:44
   6: <futures_util::future::future::Map<core::pin::Pin<alloc::boxed::Box<dyn core::future::future::Future<Output = http::response::Response<axum_core::body::Body>> + core::marker::Send>>, fn(http::response::Response<axum_core::body::Body>) -> core::result::Result<http::response::Response<axum_core::body::Body>, core::convert::Infallible>> as core::future::future::Future>::poll
             at /home/lukas/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/futures-util-0.3.31/src/lib.rs:86:35
   7: <axum::handler::future::IntoServiceFuture<core::pin::Pin<alloc::boxed::Box<dyn core::future::future::Future<Output = http::response::Response<axum_core::body::Body>> + core::marker::Send>>> as core::future::future::Future>::poll
             at /home/lukas/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/axum-0.8.8/src/macros.rs:42:39
   8: <axum::util::MapIntoResponseFuture<axum::handler::future::IntoServiceFuture<core::pin::Pin<alloc::boxed::Box<dyn core::future::future::Future<Output = http::response::Response<axum_core::body::Body>> + core::marker::Send>>>> as core::future::future::Future>::poll
             at /home/lukas/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/axum-0.8.8/src/util.rs:94:47
   9: <core::pin::Pin<alloc::boxed::Box<dyn core::future::future::Future<Output = core::result::Result<http::response::Response<axum_core::body::Body>, core::convert::Infallible>> + core::marker::Send>> as core::future::future::Future>::poll
             at /home/lukas/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/future/future.rs:133:9
  10: <tower::util::oneshot::Oneshot<tower::util::boxed_clone_sync::BoxCloneSyncService<http::request::Request<axum_core::body::Body>, http::response::Response<axum_core::body::Body>, core::convert::Infallible>, http::request::Request<axum_core::body::Body>> as core::future::future::Future>::poll
             at /home/lukas/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/tower-0.5.3/src/util/oneshot.rs:96:42
  11: <axum::routing::route::RouteFuture<core::convert::Infallible> as core::future::future::Future>::poll
             at /home/lukas/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/axum-0.8.8/src/routing/route.rs:149:41
  12: <hyper_util::service::oneshot::Oneshot<tower::util::map_request::MapRequest<axum::routing::Router, axum::serve::handle_connection<tokio::net::tcp::listener::TcpListener, axum::routing::Router, axum::routing::Router>::{closure#0}::{closure#2}>, http::request::Request<hyper::body::incoming::Incoming>> as core::future::future::Future>::poll
             at /home/lukas/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/hyper-util-0.1.19/src/service/oneshot.rs:55:42
  13: <hyper_util::service::glue::TowerToHyperServiceFuture<tower::util::map_request::MapRequest<axum::routing::Router, axum::serve::handle_connection<tokio::net::tcp::listener::TcpListener, axum::routing::Router, axum::routing::Router>::{closure#0}::{closure#2}>, http::request::Request<hyper::body::incoming::Incoming>> as core::future::future::Future>::poll
             at /home/lukas/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/hyper-util-0.1.19/src/service/glue.rs:70:31
  14: <hyper::proto::h1::dispatch::Server<hyper_util::service::glue::TowerToHyperService<tower::util::map_request::MapRequest<axum::routing::Router, axum::serve::handle_connection<tokio::net::tcp::listener::TcpListener, axum::routing::Router, axum::routing::Router>::{closure#0}::{closure#2}>>, hyper::body::incoming::Incoming> as hyper::proto::h1::dispatch::Dispatch>::poll_msg
             at /home/lukas/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/hyper-1.8.1/src/proto/h1/dispatch.rs:538:48
  15: <hyper::proto::h1::dispatch::Dispatcher<hyper::proto::h1::dispatch::Server<hyper_util::service::glue::TowerToHyperService<tower::util::map_request::MapRequest<axum::routing::Router, axum::serve::handle_connection<tokio::net::tcp::listener::TcpListener, axum::routing::Router, axum::routing::Router>::{closure#0}::{closure#2}>>, hyper::body::incoming::Incoming>, axum_core::body::Body, hyper_util::common::rewind::Rewind<hyper_util::rt::tokio::TokioIo<tokio::net::tcp::stream::TcpStream>>, hyper::proto::h1::role::Server>>::poll_write
             at /home/lukas/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/hyper-1.8.1/src/proto/h1/dispatch.rs:336:72
  16: <hyper::proto::h1::dispatch::Dispatcher<hyper::proto::h1::dispatch::Server<hyper_util::service::glue::TowerToHyperService<tower::util::map_request::MapRequest<axum::routing::Router, axum::serve::handle_connection<tokio::net::tcp::listener::TcpListener, axum::routing::Router, axum::routing::Router>::{closure#0}::{closure#2}>>, hyper::body::incoming::Incoming>, axum_core::body::Body, hyper_util::common::rewind::Rewind<hyper_util::rt::tokio::TokioIo<tokio::net::tcp::stream::TcpStream>>, hyper::proto::h1::role::Server>>::poll_loop
             at /home/lukas/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/hyper-1.8.1/src/proto/h1/dispatch.rs:173:26
  17: <hyper::proto::h1::dispatch::Dispatcher<hyper::proto::h1::dispatch::Server<hyper_util::service::glue::TowerToHyperService<tower::util::map_request::MapRequest<axum::routing::Router, axum::serve::handle_connection<tokio::net::tcp::listener::TcpListener, axum::routing::Router, axum::routing::Router>::{closure#0}::{closure#2}>>, hyper::body::incoming::Incoming>, axum_core::body::Body, hyper_util::common::rewind::Rewind<hyper_util::rt::tokio::TokioIo<tokio::net::tcp::stream::TcpStream>>, hyper::proto::h1::role::Server>>::poll_inner
             at /home/lukas/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/hyper-1.8.1/src/proto/h1/dispatch.rs:149:21
  18: <hyper::proto::h1::dispatch::Dispatcher<hyper::proto::h1::dispatch::Server<hyper_util::service::glue::TowerToHyperService<tower::util::map_request::MapRequest<axum::routing::Router, axum::serve::handle_connection<tokio::net::tcp::listener::TcpListener, axum::routing::Router, axum::routing::Router>::{closure#0}::{closure#2}>>, hyper::body::incoming::Incoming>, axum_core::body::Body, hyper_util::common::rewind::Rewind<hyper_util::rt::tokio::TokioIo<tokio::net::tcp::stream::TcpStream>>, hyper::proto::h1::role::Server>>::poll_catch
             at /home/lukas/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/hyper-1.8.1/src/proto/h1/dispatch.rs:128:33
  19: <hyper::proto::h1::dispatch::Dispatcher<hyper::proto::h1::dispatch::Server<hyper_util::service::glue::TowerToHyperService<tower::util::map_request::MapRequest<axum::routing::Router, axum::serve::handle_connection<tokio::net::tcp::listener::TcpListener, axum::routing::Router, axum::routing::Router>::{closure#0}::{closure#2}>>, hyper::body::incoming::Incoming>, axum_core::body::Body, hyper_util::common::rewind::Rewind<hyper_util::rt::tokio::TokioIo<tokio::net::tcp::stream::TcpStream>>, hyper::proto::h1::role::Server> as core::future::future::Future>::poll
             at /home/lukas/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/hyper-1.8.1/src/proto/h1/dispatch.rs:471:14
  20: <hyper::server::conn::http1::UpgradeableConnection<hyper_util::common::rewind::Rewind<hyper_util::rt::tokio::TokioIo<tokio::net::tcp::stream::TcpStream>>, hyper_util::service::glue::TowerToHyperService<tower::util::map_request::MapRequest<axum::routing::Router, axum::serve::handle_connection<tokio::net::tcp::listener::TcpListener, axum::routing::Router, axum::routing::Router>::{closure#0}::{closure#2}>>> as core::future::future::Future>::poll
             at /home/lukas/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/hyper-1.8.1/src/server/conn/http1.rs:539:51
  21: <hyper_util::server::conn::auto::UpgradeableConnection<hyper_util::rt::tokio::TokioIo<tokio::net::tcp::stream::TcpStream>, hyper_util::service::glue::TowerToHyperService<tower::util::map_request::MapRequest<axum::routing::Router, axum::serve::handle_connection<tokio::net::tcp::listener::TcpListener, axum::routing::Router, axum::routing::Router>::{closure#0}::{closure#2}>>, hyper_util::rt::tokio::TokioExecutor> as core::future::future::Future>::poll
             at /home/lukas/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/hyper-util-0.1.19/src/server/conn/auto/mod.rs:694:33
  22: <core::pin::Pin<&mut hyper_util::server::conn::auto::UpgradeableConnection<hyper_util::rt::tokio::TokioIo<tokio::net::tcp::stream::TcpStream>, hyper_util::service::glue::TowerToHyperService<tower::util::map_request::MapRequest<axum::routing::Router, axum::serve::handle_connection<tokio::net::tcp::listener::TcpListener, axum::routing::Router, axum::routing::Router>::{closure#0}::{closure#2}>>, hyper_util::rt::tokio::TokioExecutor>> as core::future::future::Future>::poll
             at /home/lukas/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/future/future.rs:133:9
  23: axum::serve::handle_connection::<tokio::net::tcp::listener::TcpListener, axum::routing::Router, axum::routing::Router>::{closure#0}::{closure#3}::{closure#0}
             at /home/lukas/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/tokio-1.49.0/src/macros/select.rs:707:49
  24: <core::future::poll_fn::PollFn<axum::serve::handle_connection<tokio::net::tcp::listener::TcpListener, axum::routing::Router, axum::routing::Router>::{closure#0}::{closure#3}::{closure#0}> as core::future::future::Future>::poll
             at /home/lukas/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/future/poll_fn.rs:151:9
  25: axum::serve::handle_connection::<tokio::net::tcp::listener::TcpListener, axum::routing::Router, axum::routing::Router>::{closure#0}::{closure#3}
             at /home/lukas/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/axum-0.8.8/src/serve/mod.rs:400:13
  26: <tokio::runtime::task::core::Core<axum::serve::handle_connection<tokio::net::tcp::listener::TcpListener, axum::routing::Router, axum::routing::Router>::{closure#0}::{closure#3}, alloc::sync::Arc<tokio::runtime::scheduler::current_thread::Handle>>>::poll::{closure#0}
             at /home/lukas/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/tokio-1.49.0/src/runtime/task/core.rs:374:24
  27: <tokio::loom::std::unsafe_cell::UnsafeCell<tokio::runtime::task::core::Stage<axum::serve::handle_connection<tokio::net::tcp::listener::TcpListener, axum::routing::Router, axum::routing::Router>::{closure#0}::{closure#3}>>>::with_mut::<core::task::poll::Poll<()>, <tokio::runtime::task::core::Core<axum::serve::handle_connection<tokio::net::tcp::listener::TcpListener, axum::routing::Router, axum::routing::Router>::{closure#0}::{closure#3}, alloc::sync::Arc<tokio::runtime::scheduler::current_thread::Handle>>>::poll::{closure#0}>
             at /home/lukas/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/tokio-1.49.0/src/loom/std/unsafe_cell.rs:16:9
  28: <tokio::runtime::task::core::Core<axum::serve::handle_connection<tokio::net::tcp::listener::TcpListener, axum::routing::Router, axum::routing::Router>::{closure#0}::{closure#3}, alloc::sync::Arc<tokio::runtime::scheduler::current_thread::Handle>>>::poll
             at /home/lukas/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/tokio-1.49.0/src/runtime/task/core.rs:363:30
  29: tokio::runtime::task::harness::poll_future::<axum::serve::handle_connection<tokio::net::tcp::listener::TcpListener, axum::routing::Router, axum::routing::Router>::{closure#0}::{closure#3}, alloc::sync::Arc<tokio::runtime::scheduler::current_thread::Handle>>::{closure#0}
             at /home/lukas/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/tokio-1.49.0/src/runtime/task/harness.rs:535:30
  30: <core::panic::unwind_safe::AssertUnwindSafe<tokio::runtime::task::harness::poll_future<axum::serve::handle_connection<tokio::net::tcp::listener::TcpListener, axum::routing::Router, axum::routing::Router>::{closure#0}::{closure#3}, alloc::sync::Arc<tokio::runtime::scheduler::current_thread::Handle>>::{closure#0}> as core::ops::function::FnOnce<()>>::call_once
             at /home/lukas/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/panic/unwind_safe.rs:274:9
  31: std::panicking::catch_unwind::do_call::<core::panic::unwind_safe::AssertUnwindSafe<tokio::runtime::task::harness::poll_future<axum::serve::handle_connection<tokio::net::tcp::listener::TcpListener, axum::routing::Router, axum::routing::Router>::{closure#0}::{closure#3}, alloc::sync::Arc<tokio::runtime::scheduler::current_thread::Handle>>::{closure#0}>, core::task::poll::Poll<()>>
             at /home/lukas/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/std/src/panicking.rs:590:40
  32: __rust_try
  33: std::panicking::catch_unwind::<core::task::poll::Poll<()>, core::panic::unwind_safe::AssertUnwindSafe<tokio::runtime::task::harness::poll_future<axum::serve::handle_connection<tokio::net::tcp::listener::TcpListener, axum::routing::Router, axum::routing::Router>::{closure#0}::{closure#3}, alloc::sync::Arc<tokio::runtime::scheduler::current_thread::Handle>>::{closure#0}>>
             at /home/lukas/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/std/src/panicking.rs:553:19
  34: std::panic::catch_unwind::<core::panic::unwind_safe::AssertUnwindSafe<tokio::runtime::task::harness::poll_future<axum::serve::handle_connection<tokio::net::tcp::listener::TcpListener, axum::routing::Router, axum::routing::Router>::{closure#0}::{closure#3}, alloc::sync::Arc<tokio::runtime::scheduler::current_thread::Handle>>::{closure#0}>, core::task::poll::Poll<()>>
             at /home/lukas/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/std/src/panic.rs:359:14
  35: tokio::runtime::task::harness::poll_future::<axum::serve::handle_connection<tokio::net::tcp::listener::TcpListener, axum::routing::Router, axum::routing::Router>::{closure#0}::{closure#3}, alloc::sync::Arc<tokio::runtime::scheduler::current_thread::Handle>>
             at /home/lukas/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/tokio-1.49.0/src/runtime/task/harness.rs:523:18
  36: <tokio::runtime::task::harness::Harness<axum::serve::handle_connection<tokio::net::tcp::listener::TcpListener, axum::routing::Router, axum::routing::Router>::{closure#0}::{closure#3}, alloc::sync::Arc<tokio::runtime::scheduler::current_thread::Handle>>>::poll_inner
             at /home/lukas/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/tokio-1.49.0/src/runtime/task/harness.rs:210:27
  37: <tokio::runtime::task::harness::Harness<axum::serve::handle_connection<tokio::net::tcp::listener::TcpListener, axum::routing::Router, axum::routing::Router>::{closure#0}::{closure#3}, alloc::sync::Arc<tokio::runtime::scheduler::current_thread::Handle>>>::poll
             at /home/lukas/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/tokio-1.49.0/src/runtime/task/harness.rs:155:20
  38: tokio::runtime::task::raw::poll::<axum::serve::handle_connection<tokio::net::tcp::listener::TcpListener, axum::routing::Router, axum::routing::Router>::{closure#0}::{closure#3}, alloc::sync::Arc<tokio::runtime::scheduler::current_thread::Handle>>
             at /home/lukas/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/tokio-1.49.0/src/runtime/task/raw.rs:337:13
  39: <tokio::runtime::task::raw::RawTask>::poll
             at /home/lukas/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/tokio-1.49.0/src/runtime/task/raw.rs:267:18
  40: <tokio::runtime::task::LocalNotified<alloc::sync::Arc<tokio::runtime::scheduler::current_thread::Handle>>>::run
             at /home/lukas/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/tokio-1.49.0/src/runtime/task/mod.rs:515:13
  41: <tokio::runtime::scheduler::current_thread::CoreGuard>::block_on::<core::pin::Pin<&mut core::pin::Pin<&mut dyn core::future::future::Future<Output = ()>>>>::{closure#0}::{closure#1}
             at /home/lukas/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/tokio-1.49.0/src/runtime/scheduler/current_thread/mod.rs:800:30
  42: tokio::task::coop::with_budget::<(), <tokio::runtime::scheduler::current_thread::CoreGuard>::block_on<core::pin::Pin<&mut core::pin::Pin<&mut dyn core::future::future::Future<Output = ()>>>>::{closure#0}::{closure#1}>
             at /home/lukas/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/tokio-1.49.0/src/task/coop/mod.rs:167:5
  43: tokio::task::coop::budget::<(), <tokio::runtime::scheduler::current_thread::CoreGuard>::block_on<core::pin::Pin<&mut core::pin::Pin<&mut dyn core::future::future::Future<Output = ()>>>>::{closure#0}::{closure#1}>
             at /home/lukas/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/tokio-1.49.0/src/task/coop/mod.rs:133:5
  44: <tokio::runtime::scheduler::current_thread::Context>::run_task::<(), <tokio::runtime::scheduler::current_thread::CoreGuard>::block_on<core::pin::Pin<&mut core::pin::Pin<&mut dyn core::future::future::Future<Output = ()>>>>::{closure#0}::{closure#1}>::{closure#0}
             at /home/lukas/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/tokio-1.49.0/src/runtime/scheduler/current_thread/mod.rs:368:43
  45: <tokio::runtime::scheduler::current_thread::Context>::enter::<(), <tokio::runtime::scheduler::current_thread::Context>::run_task<(), <tokio::runtime::scheduler::current_thread::CoreGuard>::block_on<core::pin::Pin<&mut core::pin::Pin<&mut dyn core::future::future::Future<Output = ()>>>>::{closure#0}::{closure#1}>::{closure#0}>
             at /home/lukas/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/tokio-1.49.0/src/runtime/scheduler/current_thread/mod.rs:442:19
  46: <tokio::runtime::scheduler::current_thread::Context>::run_task::<(), <tokio::runtime::scheduler::current_thread::CoreGuard>::block_on<core::pin::Pin<&mut core::pin::Pin<&mut dyn core::future::future::Future<Output = ()>>>>::{closure#0}::{closure#1}>
             at /home/lukas/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/tokio-1.49.0/src/runtime/scheduler/current_thread/mod.rs:368:28
  47: <tokio::runtime::scheduler::current_thread::CoreGuard>::block_on::<core::pin::Pin<&mut core::pin::Pin<&mut dyn core::future::future::Future<Output = ()>>>>::{closure#0}
             at /home/lukas/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/tokio-1.49.0/src/runtime/scheduler/current_thread/mod.rs:796:43
  48: <tokio::runtime::scheduler::current_thread::CoreGuard>::enter::<<tokio::runtime::scheduler::current_thread::CoreGuard>::block_on<core::pin::Pin<&mut core::pin::Pin<&mut dyn core::future::future::Future<Output = ()>>>>::{closure#0}, core::option::Option<()>>::{closure#0}
             at /home/lukas/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/tokio-1.49.0/src/runtime/scheduler/current_thread/mod.rs:840:68
  49: <tokio::runtime::context::scoped::Scoped<tokio::runtime::scheduler::Context>>::set::<<tokio::runtime::scheduler::current_thread::CoreGuard>::enter<<tokio::runtime::scheduler::current_thread::CoreGuard>::block_on<core::pin::Pin<&mut core::pin::Pin<&mut dyn core::future::future::Future<Output = ()>>>>::{closure#0}, core::option::Option<()>>::{closure#0}, (alloc::boxed::Box<tokio::runtime::scheduler::current_thread::Core>, core::option::Option<()>)>
             at /home/lukas/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/tokio-1.49.0/src/runtime/context/scoped.rs:40:9
  50: tokio::runtime::context::set_scheduler::<(alloc::boxed::Box<tokio::runtime::scheduler::current_thread::Core>, core::option::Option<()>), <tokio::runtime::scheduler::current_thread::CoreGuard>::enter<<tokio::runtime::scheduler::current_thread::CoreGuard>::block_on<core::pin::Pin<&mut core::pin::Pin<&mut dyn core::future::future::Future<Output = ()>>>>::{closure#0}, core::option::Option<()>>::{closure#0}>::{closure#0}
             at /home/lukas/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/tokio-1.49.0/src/runtime/context.rs:176:38
  51: <std::thread::local::LocalKey<tokio::runtime::context::Context>>::try_with::<tokio::runtime::context::set_scheduler<(alloc::boxed::Box<tokio::runtime::scheduler::current_thread::Core>, core::option::Option<()>), <tokio::runtime::scheduler::current_thread::CoreGuard>::enter<<tokio::runtime::scheduler::current_thread::CoreGuard>::block_on<core::pin::Pin<&mut core::pin::Pin<&mut dyn core::future::future::Future<Output = ()>>>>::{closure#0}, core::option::Option<()>>::{closure#0}>::{closure#0}, (alloc::boxed::Box<tokio::runtime::scheduler::current_thread::Core>, core::option::Option<()>)>
             at /home/lukas/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/std/src/thread/local.rs:508:12
  52: <std::thread::local::LocalKey<tokio::runtime::context::Context>>::with::<tokio::runtime::context::set_scheduler<(alloc::boxed::Box<tokio::runtime::scheduler::current_thread::Core>, core::option::Option<()>), <tokio::runtime::scheduler::current_thread::CoreGuard>::enter<<tokio::runtime::scheduler::current_thread::CoreGuard>::block_on<core::pin::Pin<&mut core::pin::Pin<&mut dyn core::future::future::Future<Output = ()>>>>::{closure#0}, core::option::Option<()>>::{closure#0}>::{closure#0}, (alloc::boxed::Box<tokio::runtime::scheduler::current_thread::Core>, core::option::Option<()>)>
             at /home/lukas/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/std/src/thread/local.rs:472:20
  53: tokio::runtime::context::set_scheduler::<(alloc::boxed::Box<tokio::runtime::scheduler::current_thread::Core>, core::option::Option<()>), <tokio::runtime::scheduler::current_thread::CoreGuard>::enter<<tokio::runtime::scheduler::current_thread::CoreGuard>::block_on<core::pin::Pin<&mut core::pin::Pin<&mut dyn core::future::future::Future<Output = ()>>>>::{closure#0}, core::option::Option<()>>::{closure#0}>
             at /home/lukas/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/tokio-1.49.0/src/runtime/context.rs:176:17
  54: <tokio::runtime::scheduler::current_thread::CoreGuard>::enter::<<tokio::runtime::scheduler::current_thread::CoreGuard>::block_on<core::pin::Pin<&mut core::pin::Pin<&mut dyn core::future::future::Future<Output = ()>>>>::{closure#0}, core::option::Option<()>>
             at /home/lukas/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/tokio-1.49.0/src/runtime/scheduler/current_thread/mod.rs:840:27
  55: <tokio::runtime::scheduler::current_thread::CoreGuard>::block_on::<core::pin::Pin<&mut core::pin::Pin<&mut dyn core::future::future::Future<Output = ()>>>>
             at /home/lukas/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/tokio-1.49.0/src/runtime/scheduler/current_thread/mod.rs:740:24
  56: <tokio::runtime::scheduler::current_thread::CurrentThread>::block_on::<core::pin::Pin<&mut dyn core::future::future::Future<Output = ()>>>::{closure#0}
             at /home/lukas/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/tokio-1.49.0/src/runtime/scheduler/current_thread/mod.rs:200:33
  57: tokio::runtime::context::runtime::enter_runtime::<<tokio::runtime::scheduler::current_thread::CurrentThread>::block_on<core::pin::Pin<&mut dyn core::future::future::Future<Output = ()>>>::{closure#0}, ()>
             at /home/lukas/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/tokio-1.49.0/src/runtime/context/runtime.rs:65:16
  58: <tokio::runtime::scheduler::current_thread::CurrentThread>::block_on::<core::pin::Pin<&mut dyn core::future::future::Future<Output = ()>>>
             at /home/lukas/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/tokio-1.49.0/src/runtime/scheduler/current_thread/mod.rs:188:9
  59: <tokio::runtime::runtime::Runtime>::block_on_inner::<core::pin::Pin<&mut dyn core::future::future::Future<Output = ()>>>
             at /home/lukas/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/tokio-1.49.0/src/runtime/runtime.rs:368:52
  60: <tokio::runtime::runtime::Runtime>::block_on::<core::pin::Pin<&mut dyn core::future::future::Future<Output = ()>>>
             at /home/lukas/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/tokio-1.49.0/src/runtime/runtime.rs:342:18
  61: exex::tests::test
             at ./src/lib.rs:22:65
  62: exex::tests::test::{closure#0}
             at ./src/lib.rs:18:17
  63: <exex::tests::test::{closure#0} as core::ops::function::FnOnce<()>>::call_once
             at /home/lukas/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/ops/function.rs:250:5
  64: core::ops::function::FnOnce::call_once
             at /rustc/ded5c06cf21d2b93bffd5d884aa6e96934ee4234/library/core/src/ops/function.rs:250:5
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
```

</p>
</details>

### Related Issues

- https://github.com/taiki-e/cargo-llvm-cov/issues/459#issuecomment-3698492302

**Comment by jyn514** [comment]
It would be helpful to see the LLVM IR for this function when compiled with -C instrument-coverage. Additionally, if you could check the .profdata files with llvm-profdata to see if they have any mentions of this function, that would be quite helpful.

**Comment by djwitte** [comment]
Hit this issue as well, and did some investigation as requested. Happy to pull more artifacts if helpful (see end).

`llvm-cov export` on the below smaller repro shows 9 functions in the coverage map; profdata holds 8. The extra one is `bad::{closure#0}::{closure#0}::{closure#0}::<_>` (mangled `_RINCNCNCNv...3bad000pEB8_`). It carries 11 source regions covering lines L34-L38 of the closure body, all with count=0. The matching real instance `bad::{closure#0}::{closure#0}` is in profdata with count=1, and its region list is the two closure-brace ranges only — none of the body lines.

### Smaller repro

One dependency (`futures-lite`). Calling the closure inline (`let f = async || { ... }; f().await`) does not trigger the bug; passing it through a generic `AsyncFnOnce` parameter does.

`Cargo.toml`:
```toml
[package]
name = "async-closure-cov-repro"
version = "0.0.0"
edition = "2024"

[dependencies]
futures-lite = "2"
```

`src/lib.rs`:
```rust
use std::sync::atomic::{AtomicUsize, Ordering};

pub static GOOD_STEPS: AtomicUsize = AtomicUsize::new(0);
pub static BAD_STEPS: AtomicUsize = AtomicUsize::new(0);

pub async fn good() {
    GOOD_STEPS.fetch_add(1, Ordering::SeqCst);
    GOOD_STEPS.fetch_add(1, Ordering::SeqCst);
    GOOD_STEPS.fetch_add(1, Ordering::SeqCst);
}

async fn call_it<F: AsyncFnOnce()>(f: F) {
    f().await
}

pub async fn bad() {
    let f = async || {
        BAD_STEPS.fetch_add(1, Ordering::SeqCst);
        BAD_STEPS.fetch_add(1, Ordering::SeqCst);
        BAD_STEPS.fetch_add(1, Ordering::SeqCst);
    };
    call_it(f).await
}

#[cfg(test)]
mod tests {
    use super::*;
    use futures_lite::future::block_on;

    #[test]
    fn both_bodies_execute_three_times() {
        block_on(good());
        block_on(bad());
        assert_eq!(GOOD_STEPS.load(Ordering::SeqCst), 3);
        assert_eq!(BAD_STEPS.load(Ordering::SeqCst), 3);
    }
}
```

Ran with `cargo +nightly llvm-cov --html test`. The test passes — both `assert_eq!(*_STEPS.load(...), 3)` succeed — but the HTML report shows `BAD_STEPS.fetch_add(...)` lines L35-L37 in red with count=0, while the matching `GOOD_STEPS` lines L24-L26 are green with count=1.

### `llvm-profdata show --all-functions --counts` (demangled)

```
async_closure_cov_repro::good                                          Counters: 1  Function count: 1
async_closure_cov_repro::good::{closure#0}                             Counters: 1  Function count: 1
async_closure_cov_repro::bad                                           Counters: 1  Function count: 1
async_closure_cov_repro::bad::{closure#0}                              Counters: 3  Function count: 1
async_closure_cov_repro::bad::{closure#0}::{closure#0}                 Counters: 1  Function count: 1
async_closure_cov_repro::call_it::<...>                                Counters: 1  Function count: 1
async_closure_cov_repro::call_it::<...>::{closure#0}                   Counters: 3  Function count: 1
async_closure_cov_repro::tests::both_bodies_execute_three_times        Counters: 3  Function count: 1
Total functions: 8
```

### `llvm-cov export` (region map, closure-related entries)

```
async_closure_cov_repro::bad::{closure#0}                              count=1
  L33:20-21  count=1  Code
  L34:9-10   count=1  Code
  L39:5-12   count=1  Code
  L39:5-21   count=1  Code
  L40:1-2    count=1  Code

async_closure_cov_repro::bad::{closure#0}::{closure#0}                 count=1
  L34:22-23  count=1  Code
  L38:5-6    count=1  Code

async_closure_cov_repro::bad::{closure#0}::{closure#0}::{closure#0}::<_>   count=0
  L34:22-23  count=0  Code
  L35:9-18   count=0  Code
  L35:19-28  count=0  Code
  L35:32-48  count=0  Code
  L36:9-18   count=0  Code
  L36:19-28  count=0  Code
  L36:32-48  count=0  Code
  L37:9-18   count=0  Code
  L37:19-28  count=0  Code
  L37:32-48  count=0  Code
  L38:5-6    count=0  Code
```

Grepping the emitted LLVM IR for the exact placeholder symbol `_RINCNCNCNv...3bad000pEB8_` finds zero matches anywhere — no `define`, no `@__profc_` global, no other reference. The IR does contain monomorphized instantiations of the same coroutine path with a concrete type substitution (`...3bad000l...`, which `rustfilt` decodes as `bad::{closure#0}::{closure#0}::{closure#0}<i32>`), but those appear only as `core::ptr::drop_glue::<...>`, `core::pin::Pin::<...>::new_unchecked`, and `<... as IntoFuture>::into_future` — never as a poll function with body code.

### LLVM IR

Built with `RUSTFLAGS="-C instrument-coverage --emit=llvm-ir" cargo +nightly build --tests`.

`good::{closure#0}` has one `atomicrmw add` against its profile counter:
```llvm
%2 = atomicrmw add ptr @__profc__RNCNvCs..._4good0B3_, i64 1 monotonic, align 8
```

`bad::{closure#0}::{closure#0}` has one `atomicrmw add` against its profile counter, immediately preceding the three `invoke @<core::sync::atomic::Atomic<usize>>::fetch_add` calls that correspond to source lines L35-L37:
```llvm
bb1:
  %2 = atomicrmw add ptr @__profc__RNsNCNCNvCs..._3bad000B7_, i64 1 monotonic, align 8
  %_3 = invoke i64 @<core::sync::atomic::Atomic<usize>>::fetch_add(...)   ; !dbg !DILocation(line: 35, column: 19, ...)
        to label %bb2 unwind label %cleanup
bb2:
  %_6 = invoke i64 @<core::sync::atomic::Atomic<usize>>::fetch_add(...)   ; !dbg !DILocation(line: 36, column: 19, ...)
        to label %bb3 unwind label %cleanup
bb3:
  %_9 = invoke i64 @<core::sync::atomic::Atomic<usize>>::fetch_add(...)   ; !dbg !DILocation(line: 37, column: 19, ...)
        to label %bb4 unwind label %cleanup
bb4:
  store i8 1, ptr %_1, align 1
  ret i1 false
```

Note the `!DILocation` debug metadata: the body code that runs is *inside* `bad::{closure#0}::{closure#0}` (the depth-2 instance that profdata records as count=1), and rustc has correctly tagged those instructions as originating from L35, L36, L37. The coverage region map just doesn't claim those lines for this function — it claims them for the depth-3 `::<_>` symbol that has no body anywhere.

### Directional hypothesis (speculative, not verified)

The mangled name `_RINCNCNCNv...3bad000pEB8_` decodes as a v0 instance with a `p` (placeholder) substitution, which `rustfilt` renders as `::<_>`. The real instance is one nesting level shallower and has no placeholder. Pattern is consistent with the coverage map being populated against a less-substituted `Instance` than the one MIR codegen actually emits, so body regions end up on a symbol that never reaches `__profc_`. Haven't checked the rustc pass that does this — flagging in case it narrows the search.

### Toolchain

```
rustc 1.98.0-nightly (23a3312d9 2026-05-23)
host: aarch64-apple-darwin
LLVM version: 22.1.6
cargo-llvm-cov 0.8.7
```

### Artifacts

[coverage-map.json](https://github.com/user-attachments/files/28187074/coverage-map.json)
[profdata-demangled.txt](https://github.com/user-attachments/files/28187075/profdata-demangled.txt)
[profdata-mangled.txt](https://github.com/user-attachments/files/28187076/profdata-mangled.txt)
[async-closure-cov-151135-artifacts.zip](https://github.com/user-attachments/files/28187091/async-closure-cov-151135-artifacts.zip)
